### PR TITLE
Reconcile vpc endpoint

### PR DESCRIPTION
**Pull request by nprokopic**
This PR:

- adds/changes/removes etc

### Testing

Description on how aws-vpc-operator can be tested.

- [ ] fresh install works
- [ ] upgrade from previous version works
- [ ] aws-vpc-operator manages only VPCs that are not managed by CAPA

#### Other testing

Description of features to additionally test for aws-vpc-operator installations.

- [ ] check reconciliation of existing resources after upgrading
- [ ] AWS VPC works after aws-vpc-operator upgrade

### Checklist

- [ ] Update changelog in CHANGELOG.md.
- [ ] Make sure `values.yaml` and `values.schema.json` are valid.

### CODE REVIEW

File: controllers/awscluster_controller.go [hunk 1]
```diff
@@ -477,16 +499,109 @@ func (r *AWSClusterReconciler) reconcileNormal(ctx context.Context, logger logr.
...
+				Spec: vpcendpoint.Spec{
+					VpcId: awsCluster.Spec.NetworkSpec.VPC.ID,
+				},
+				AdditionalTags: awsCluster.Spec.AdditionalTags,
+			},
+		}
+		for _, subnet := range awsCluster.Spec.NetworkSpec.Subnets {
+			reconcileRequest.Spec.SubnetIds = append(reconcileRequest.Spec.SubnetIds, subnet.ID)
+		}
+		for _, securityGroup := range awsCluster.Status.Network.SecurityGroups {
```
Comment (calvix) [inline controllers/awscluster_controller.go:525]: I actually saw a bug where it added only the default security group to the VPC endpoint, Might be a race condition issue

I think we should double check that there is expected amount of security group here and not only 1

Comment (calvix) [reply controllers/awscluster_controller.go:525]: we need `mario0003-node` and `mario0003-controlplane` security group for sure, so maybe check that these two are present? if not wait for another reconciliation (1 min should be ok)

Comment (nprokopic) [reply controllers/awscluster_controller.go:525]: The current logic is to wait that CAPA sets `ClusterSecurityGroupsReady` condition to true. Maybe that is not enough. I will look into what you suggested above and check again how CAPA sets security groups and how/when to be sure that everything we need is set there so we can use it.

Comment (nprokopic) [reply controllers/awscluster_controller.go:525]: These 4:
```
var defaultAWSSecurityGroupRoles = []infrav1.SecurityGroupRole{
	infrav1.SecurityGroupAPIServerLB,
	infrav1.SecurityGroupLB,
	infrav1.SecurityGroupControlPlane,
	infrav1.SecurityGroupNode,
}
```

Plus optionally bastion SG.

Comment (calvix) [reply controllers/awscluster_controller.go:525]: yep

Comment (nprokopic) [reply controllers/awscluster_controller.go:525]: One thing that occurred to me, security group IDs are taken from AWSCluster.Status, while everything else is read from the AWSCluster.Spec. So maybe Status gets updated few times, and we just pick up some initial state (we are not updating VPC endpoint after initial creation, that is one of the TODOs/missing things, so implementing updates on reconciliation could make it work)

Comment (calvix) [reply controllers/awscluster_controller.go:525]: hit the same issue again, so its not very rare and needs to be adresset

**Pull request merged**